demos/racetrack/Driver.cpp

```diff
@@ -83,7 +102,13 @@ void MyFrameHandler(unsigned long windowID, unsigned int viewport, void *)
 	// Draw map
 	r->Draw(display);
 	// Draw "racecar"
-	r->Draw(display, s);
+	r->Draw(display, s); //Draws the state of the racetrack
+	r->GoalTest(s, s);
+	if (move == true) // if m was pressed, makes the agent move automatically
+	{
+		r->ApplyAction(s, v);
+	}
+	Boundaries(); // doesn't let the agent out of the map
```

Comment:
This check should be moved into GetActions and ApplyActions -- the main Driver code shouldn't know the details of how boundaries are checked. For instance, in other environments we often create a method called Legal:

`	bool Legal(const state &node1, const action &act) const;`

You can use this to check if you are applying a legal action.